src/main.rs

```diff
@@ -1,126 +1,267 @@
+    view: Mat4<f32>,
+    projection: Mat4<f32>,
+    frames: &[Vec<Mesh>],
+) {
+    let rows = 2;
+    let columns = 4;
+    assert!(frames.len() <= rows * columns, "not enough room on spritesheet for all sprites");
+
+    let mut img = ImageBuffer::new((image_width * columns) as u32, (image_height * rows) as u32);
+
+    let mut color = Buffer2d::new([image_width, image_height], 0);
+    let mut depth = Buffer2d::new([image_width, image_height], 1.0);
+
+    for (i, frame) in frames.into_iter().enumerate() {
+        render(&mut color, &mut depth, model, view, projection, frame, 0.15);
+
+        let column = i % columns;
+        let row = i / columns;
+
+        for x in 0..image_width {
+            for y in 0..image_height {
+                // Unsafe because we are guaranteeing that these indexes are not out of bounds
+                let color = unsafe { color.get([x, y]) };
+                let rgba = bgra_u32_to_rgba(*color);
+
+                let pixel = img.get_pixel_mut(
+                    (column * image_width + x) as u32,
+                    (row * image_height + y) as u32,
+                );
+
+                *pixel = vek_rgba_to_image_rgba(rgba);
+            }
+        }
+
+        color.clear(0);
+        depth.clear(1.0);
     }
-}
 
-//TODO: Read the modules chapter and split this file up
-
-/// The ID of a vertex within a specific mesh
-/// Only to be used for the mesh where it came from
-#[derive(Debug, Clone)]
-struct VertexId(usize);
-/// The ID of a edge within a specific mesh
-/// Only to be used for the mesh where it came from
-#[derive(Debug, Clone)]
-struct EdgeId(usize);
-/// The ID of a face within a specific mesh
-/// Only to be used for the mesh where it came from
-#[derive(Debug, Clone)]
-struct FaceId(usize); //TODO: lookup "rust newtype pattern"
-//TODO: Add good docs to everything
-#[derive(Debug, Clone)]
-struct MeshId(usize); //TODO: index into scene mesh array?
-
-#[derive(Debug, Clone)]
-struct Mesh {
-    // Used to associate with other things
-    id: MeshId, //TODO: Figure out the right type for this
-    vertices: Vec<Point2D>,
-    /// List of edge indexes (a, b) == (b, a)
-    edges: Vec<(VertexId, VertexId)>,
-    faces: Vec<Vec<VertexId>>,
-    //TODO: Associate vertices with material IDs
+    img.save("spritesheet.png").expect("unable to write image");
 }
 
-impl Mesh {
-    pub fn vertex(&self, vid: VertexId) -> Point2D {
-        unimplemented!();
-    }
+fn preview_window(
+    image_width: usize,
+    image_height: usize,
+    model: Mat4<f32>,
+    view: Mat4<f32>,
+    projection: Mat4<f32>,
+    frames: &[Vec<Mesh>],
+) {
+    let scale = 16;
+    let background = rgba_to_bgra_u32(Rgba {r: 0.62, g: 0.62, b: 0.62, a: 1.0});
 
-    pub fn edge(&self, eid: EdgeId) -> (Point2D, Point2D) {
-        unimplemented!();
-    }
+    let mut color = Buffer2d::new([image_width, image_height], background);
+    let mut depth = Buffer2d::new([image_width, image_height], 1.0);
```

Comment:
It seems like we have to initialize a depth buffer of the same size twice (once here, and once in save_spritesheet). Would it be beneficial to create a persistent depth buffer that we only need to initialize once?

Reply:
If this can be done, and has benefits, we can do this to the color buffer too, and add a mutator to the Buffer2d type to set the color?

Reply:
Persistent depth buffer is probably not a good idea. We want these functions and their buffers to remain separate. These will eventually be completely separate parts of the app. If we just want to avoid duplicating the initialization logic we should consider having a wrapper struct that maintains the buffers separately. Definitely a lot of improvements that can be made to this prototype!

Reply:
I'm not entirely sure why we want to have separate buffers if we don't require to generate both the sprite sheet and the animation window at the same time, but maybe we can talk about this more in a future meeting :) Also, the amount of memory we're using is small enough that it virtually doesn't matter so I'll close this issue.